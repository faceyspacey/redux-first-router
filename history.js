module.exports = {
  createHistory: require('./dist/core/createHistory.js').default,
  History: require('./dist/history/History.js').default,
  MemoryHistory: require('./core/history/MemoryHistory.js').default,
  BrowserHistory: require('./core/history/BrowserHistory.js').default,
  utils: {
    supports: require('./dist/history/utils/supports.js'),
    popListener: require('./core/history/utils/popListener.js'),
    sessionStorage: require('./core/history/utils/sessionStorage.js')
  }
}


/** if you want to extend History, here is how you do it:

import History from 'rudy/history/BrowserHistory'

class MyHistory extends History {
  push(path) {
    const location = this.createAction(path)
    // do something custom
  }
}

// usage:

import { createRouter } from 'rudy'
import { createHistory as creatHist } from 'rudy/core'

const createHistory = (routes, opts) => {
  if (opts.someCondition) return new MyHistory(routes, opts)
  return creatHist(routes, opts)
}

const { middleware, reducer, firstRoute } = createRouter(routes, { createHistory })

*/
