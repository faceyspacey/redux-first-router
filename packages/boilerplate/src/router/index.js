/* eslint-disable */
import React, { Component } from 'react'
import { connect } from 'react-redux'
import * as Rudy from '@respond-framework/rudy'

const routesMap = {}

class Router extends Component {
  constructor(props) {
    super(props);

    const childs = Array.isArray(props.children) ? props.children : [props.children]

    const routeKey = childs[0].props.component().chunkName()
    const routes = {
      [routeKey]: {
        ...childs[0].props,
      },
    }
    const formatRoute = (route) => ({
      ...route,
    })
    props.addRoutes(routes, formatRoute)
  }
  render() {

    // this.props.setRoute({ type: routeKey })
    return <span>route</span>
  }
}

class Route extends Component {
  render() {
    return null
  }
}

const getRoutesMap = () => routesMap


const mapDispatchToProps = (dispatch) => ({
  addRoutes: (routes, formatRoute = null) => dispatch(Rudy.addRoutes(routes)),
  setRoute: (object) => dispatch(object)
})

const mapState = (state) => {
  return {}
}

const ConnectedRouter = connect(mapState, mapDispatchToProps)(Router)

export {
  Route,
  ConnectedRouter,
  getRoutesMap
}
