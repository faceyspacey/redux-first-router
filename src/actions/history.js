// @flow
import { CALL_HISTORY } from '../types'
import type { HistoryRouteAction } from '../flow-types'


export const push = (path: string, state: ?Object): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'push',
    args: [path, state]
  }
})

export const replace = (path: string, state: ?Object): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'replace',
    args: [path, state]
  }
})


export const jump = (delta: number | string, state: ?Object, byIndex: ?boolean, n: ?number): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'jump',
    args: [delta, state, byIndex, n]
  }
})

export const reset = (entries: Array<string | Object>, index: ?number, n: ?number): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'reset',
    args: [entries, index, n]
  }
})

export const back = (state: ?(Object | Function)): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'back',
    args: [state]
  }
})

export const next = (state: ?(Object | Function)): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'next',
    args: [state]
  }
})

export const set = (action: Object | Function, n: ?(number | string), byIndex: ?boolean): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'set',
    args: [action, n, byIndex]
  }
})

export const setParams = (params: Object | Function, n: ?(number | string), byIndex: ?boolean): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'set',
    args: [{ params }, n, byIndex]
  }
})

export const setQuery = (query: Object | Function, n: ?(number | string), byIndex: ?boolean): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'set',
    args: [{ query }, n, byIndex]
  }
})

export const setState = (state: Object | Function, n: ?(number | string), byIndex: ?boolean): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'set',
    args: [{ state }, n, byIndex]
  }
})

export const setHash = (hash: Object | Function, n: ?(number | string), byIndex: ?boolean): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'set',
    args: [{ hash }, n, byIndex]
  }
})

export const setBasename = (basename: Object | Function, n: ?(number | string), byIndex: ?boolean): HistoryRouteAction => ({
  type: CALL_HISTORY,
  payload: {
    method: 'set',
    args: [{ basename }, n, byIndex]
  }
})

// NOTE: see `src/utils/formatRoutes.js` for implementation of corresponding pathlessRoute
