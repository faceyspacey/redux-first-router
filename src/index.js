// @flow

export { default as createRouter } from './core/createRouter'
export { default as createScene } from './createScene'

export type {
  RouteInput,
  Route,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
  Navigators,
  Navigator,
  Store,
  NavigationAction,
  Router
} from './flow-types'
