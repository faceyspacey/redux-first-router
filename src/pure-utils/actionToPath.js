// @flow
import pathToRegexp from 'path-to-regexp'
import type {
  RouteObject,
  Payload,
  Params,
  RoutesMap,
  ReceivedAction as Action,
  QuerySerializer
} from '../flow-types'

export default (
  action: Action,
  routesMap: RoutesMap,
  serializer?: QuerySerializer
): string => {
  const route = routesMap[action.type]
  const routePath = typeof route === 'object' ? route.path : route
  const toPathOptions = typeof route === 'object'
    ? route.toPathOptions
    : undefined
  const params = typeof route === 'object'
    ? _payloadToParams(route, action.payload)
    : action.payload

  const path =
    pathToRegexp.compile(routePath)(params || {}, toPathOptions) || '/'

  const query =
    action.query ||
    (action.meta && action.meta.query) ||
    (action.payload && action.payload.query)

  const search = query && serializer && serializer.stringify(query)

  return search ? `${path}?${search}` : path
}

const _payloadToParams = (route: RouteObject, params: Payload = {}): Params =>
  Object.keys(params).reduce((sluggifedParams, key) => {
    if (typeof params[key] !== 'undefined') {
      if (typeof params[key] === 'number') {
        sluggifedParams[key] = params[key]
      }
      else if (route.capitalizedWords === true) {
        sluggifedParams[key] = params[key].replace(/ /g, '-').toLowerCase()
      }
      else if (typeof route.toPath === 'function') {
        sluggifedParams[key] = route.toPath(params[key], key)
      }
      else if (typeof params[key] === 'string') {
        sluggifedParams[key] = params[key]
      }
    }

    return sluggifedParams
  }, {})
