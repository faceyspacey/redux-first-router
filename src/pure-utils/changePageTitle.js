// @flow
import type { Document } from '../flow-types'

export default (doc: Document, title: ?string): ?string => {
  if (typeof title === 'string' && doc.title !== title) {
    return (doc.title = title)
  }

  return null
}

export const getDocument = (): Document => {
  const isSSRTest = process.env.NODE_ENV === 'test' && typeof window !== 'undefined' && window.isSSR

  return typeof document !== 'undefined' && !isSSRTest ? document : {}
}
