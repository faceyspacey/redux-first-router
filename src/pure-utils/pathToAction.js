// @flow
import { compilePath } from 'rudy-match-path'
import { stripBasename } from 'rudy-history/PathUtils'
import { NOT_FOUND, getOptions } from '../index'
import objectValues from './objectValues'

import type { RoutesMap, ReceivedAction, QuerySerializer } from '../flow-types'

export default (
  pathname: string,
  routesMap: RoutesMap,
  serializer?: QuerySerializer,
  basename?: string | void = getOptions().basename,
  strict?: boolean | void = getOptions().strict
): ReceivedAction => {
  const parts = pathname.split('?')
  const search = parts[1]
  const query = search && serializer && serializer.parse(search)
  const routes = objectValues(routesMap)
  const routeTypes = Object.keys(routesMap)

  pathname = basename ? stripBasename(parts[0], basename) : parts[0]

  let i = 0
  let match
  let keys

  while (!match && i < routes.length) {
    const regPath = typeof routes[i] === 'string' ? routes[i] : routes[i].path // route may be an object containing a route or a route string itself

    if (!regPath) {
      i++
      continue
    }

    const { re, keys: k } = compilePath(regPath, { strict })
    match = re.exec(pathname)
    keys = k
    i++
  }

  if (match) {
    i--

    const capitalizedWords =
      typeof routes[i] === 'object' && routes[i].capitalizedWords


    const coerceNumbers =
      typeof routes[i] === 'object' && routes[i].coerceNumbers

    const fromPath =
      routes[i] &&
      typeof routes[i].fromPath === 'function' &&
      routes[i].fromPath

    const userMeta = typeof routes[i] === 'object' && routes[i].meta

    const type = routeTypes[i]

    const payload = (keys || []).reduce((payload, key, index) => {
      let val = match && match[index + 1] // item at index 0 is the overall match, whereas those after correspond to the key's index

      if (typeof val === 'string') {
        if (fromPath) {
          val = fromPath && fromPath(val, key.name)
        }
        else if (coerceNumbers && isNumber(val)) {
          val = parseFloat(val)
        }
        else if (capitalizedWords) {
          val = val.replace(/-/g, ' ').replace(/\b\w/g, l => l.toUpperCase()) // 'my-category' -> 'My Category'
        }
      }

      payload[key.name] = val
      return payload
    }, {})

    const meta = {
      ...(userMeta ? { meta: userMeta } : {}),
      ...(query ? { query } : {})
    }
    return { type, payload, meta }
  }

  // This will basically will only end up being called if the developer is manually calling history.push().
  // Or, if visitors visit an invalid URL, the developer can use the NOT_FOUND type to show a not-found page to
  const meta = { notFoundPath: pathname, ...(query ? { query } : {}) }
  return { type: NOT_FOUND, payload: {}, meta }
}

const isNumber = (val: string) => /^\d+$/.test(val)
