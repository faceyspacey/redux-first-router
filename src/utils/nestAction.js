// @flow
import { isNotFound } from './index'
import type { ActionToNavigation, LocationState } from '../flow-types'

export default (action: ActionToNavigation, prevState: LocationState, fromAction: LocationState, statusCode: number, tmp: Object = {}) => {
  const { location, type, basename, params, query, state, hash } = action
  const { entries, index, length, pathname, search, url, key, scene, n } = location

  const prev = createActionRef(prevState)
  const from = createActionRef(fromAction)

  const kind = resolveKind(location.kind, tmp.load, from)
  const direction = n === -1 ? 'backward' : 'forward'

  const pop = !!tmp.revertPop
  const status = from ? (statusCode || 302) : (isNotFound(type) ? 404 : 200)

  return {
    type,
    params,
    query,
    state,
    hash,
    basename,
    location: {
      kind,
      direction,
      n,

      url,
      pathname,
      search,
      key,
      scene,

      prev,
      from,
      blocked: null,

      entries,
      index,
      length,

      pop,
      status
    }
  }
}

export const createActionRef = (actionOrState: Object) => {
  if (!actionOrState) return null

  // if `prev` or redirect action from outside of pipeline, we receive the state instead (see ./formatAction.js)
  if (!actionOrState.location) {
    const { type, params, query, state, hash, basename, ...rest } = actionOrState
    const location = createLocationRef(rest)
    const action = { type, params, query, state, hash, basename, location }
    return action
  }

  // if redirect occurred during pipeline, we receive an action representing the previous state
  return {
    ...actionOrState,
    location: createLocationRef({ ...actionOrState.location })
  }
}

const createLocationRef = (loc) => {
  delete loc.prev
  delete loc.from
  delete loc.blocked
  delete loc.universal

  delete loc.length
  delete loc.kind
  delete loc.entries
  delete loc.pop
  delete loc.status
  delete loc.direction
  delete loc.n
  delete loc.universal
  delete loc.ready

  return loc
}

const resolveKind = (kind, isLoad, from) =>
  isLoad
    ? 'load' // insure redirects don't change kind on load
    : !from
      ? kind // PRIMARY USE CASE: preverse the standard kind
      : kind.replace('push', 'replace') // pipeline redirects before enter are in fact pushes, but users shouldn't have to think about that -- using `kind.replace` preserves back/next kinds
