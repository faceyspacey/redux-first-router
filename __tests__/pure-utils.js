import { createMemoryHistory, createBrowserHistory } from 'history'

import isLocationAction from '../src/pure-utils/isLocationAction'
import isServer from '../src/pure-utils/isServer'
import objectValues from '../src/pure-utils/objectValues'
import nestAction, { nestHistory } from '../src/pure-utils/nestAction'
import pathToAction from '../src/pure-utils/pathToAction'
import actionToPath from '../src/pure-utils/actionToPath'
import changePageTitle from '../src/pure-utils/changePageTitle'

import { NOT_FOUND } from '../src/index'

it('isLocationAction(action) if has meta.location object', () => {
  let ret = isLocationAction({})
  expect(ret).toBeFalsy()

  ret = isLocationAction({ meta: { location: { current: {} } } })
  expect(ret).toBeTruthy()
})

it('isServer()', () => {
  expect(isServer()).toEqual(false)

  global.window.SSRtest = true
  expect(isServer()).toEqual(true)
  delete global.window.SSRtest
})

it('objectValues(routesMap) converts map of routes to an array of routes without the action type keys', () => {
  const routesMap = {
    ACTION_TYPE: '/foo/:bar',
    ACTION_TYPE_2: { path: '/path/:baz/', capitalizedWords: true }
  }

  const ret = objectValues(routesMap)
  expect(ret).toEqual([routesMap.ACTION_TYPE, routesMap.ACTION_TYPE_2]) /*? */
})

describe('nestAction(pathname, receivedAction, prevLocation, history, kind?)', () => {
  it('nestAction properly formats/nests action object', () => {
    const history = createMemoryHistory()
    const pathname = '/path'
    const receivedAction = {
      type: 'FOO',
      payload: { bar: 'baz' },
      meta: { info: 'something' }
    }
    const location = {
      pathname: 'previous',
      type: 'PREV',
      payload: { bla: 'prev' }
    }

    let action = nestAction(
      pathname,
      receivedAction,
      location,
      history
    ) /*? $.meta */

    expect(action.type).toEqual('FOO')
    expect(action.payload).toEqual({ bar: 'baz' })

    expect(action.type).toEqual(action.meta.location.current.type)
    expect(action.payload).toEqual(action.meta.location.current.payload)

    expect(action.meta.location.prev).toEqual(location)
    expect(action.meta).toMatchObject(receivedAction.meta)
    expect(action.meta.location.current.pathname).toEqual(pathname)

    expect(action).toMatchSnapshot()

    expect(action.meta.location.kind).not.toBeDefined()
    action = nestAction(pathname, receivedAction, location, history, 'load')
    expect(action.meta.location.kind).toEqual('load')

    action = nestAction(pathname, receivedAction, location, history, 'pop')
    expect(action.meta.location.kind).toEqual('pop')
  })

  it('nestHistory formats simplified history object for action + state', () => {
    const history = createMemoryHistory() // still use `createMemoryHistory` for stability during tests
    history.push('/foo')
    history.push('/bar/baz')

    const nestedHistory = nestHistory(history) /*? */
    expect(nestedHistory).toMatchSnapshot()
  })

  it('nestHistory returns undefined when using createBrowserHistory', () => {
    const history = createBrowserHistory()
    const nestedHistory = nestHistory(history)
    expect(nestedHistory).toEqual(undefined)
  })
})

describe('pathToAction(path, routesMap)', () => {
  it('parse path into action using routePath without /:param segment', () => {
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param'
    }

    const action = pathToAction('/info', routesMap)
    expect(action).toEqual({ type: 'INFO', payload: {}, meta: {} }) /*? */
  })

  it('parse path into action using routePath with /:param segment', () => {
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param'
    }

    const action = pathToAction('/info/foo', routesMap)
    expect(action).toEqual({
      type: 'INFO_PARAM',
      payload: { param: 'foo' },
      meta: {}
    }) /*? */
  })

  it('parse path (/info/foo-bar) into action using route object containing capitalizedWords: true: payload: { param: "Foo Bar" }', () => {
    const path = '/info/foo-bar'
    const routesMap = {
      INFO_PARAM: { path: '/info/:param', capitalizedWords: true }
    }

    const action = pathToAction(path, routesMap) /*? */
    expect(action.payload.param).toEqual('Foo Bar')
  })

  it('parse path into action using route object containing fromPath() function', () => {
    const path = '/info/foo-bar/'
    const routesMap = {
      INFO_PARAM: {
        path: '/info/:param',
        fromPath: (segment, key) =>
          `${segment} ${key}`.replace('-', ' ').toUpperCase()
      }
    }

    const action = pathToAction(path, routesMap) /*? */
    expect(action.payload.param).toEqual('FOO BAR PARAM')
  })

  it('parse path containing number param into action with payload value set as integer instead of string', () => {
    const path = '/info/69/'
    const routesMap = {
      INFO_PARAM: { path: '/info/:param' }
    }

    const action = pathToAction(path, routesMap) /*? */
    expect(typeof action.payload.param).toEqual('number')
    expect(action.payload.param).toEqual(69)
  })

  it('does not parse a blank string "" as NaN', () => {
    const path = '/info'
    const routesMap = {
      INFO_WILDCARD: { path: '/info(.*)' }
    }

    const action = pathToAction(path, routesMap)
    expect(action.payload[0]).toEqual('')
  })

  it('parsed path not found and return NOT_FOUND action.type: "@@redux-first-router/NOT_FOUND"', () => {
    const path = '/info/foo/bar'
    const routesMap = {
      INFO_PARAM: { path: '/info/:param/' }
    }

    const action = pathToAction(path, routesMap) /*? */
    expect(action.type).toEqual(NOT_FOUND)
  })
})

describe('actionToPath(action, routesMap)', () => {
  it('parse action into path without payload: /info', () => {
    const action = { type: 'INFO' }
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param'
    }

    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/info')
  })

  it('parse action payload into path segment: /info/foo', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'foo' } }
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param'
    }

    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/info/foo')
  })

  it('parse action into path with numerical payload key value: /info/69', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 69 } }
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: { path: '/info/:param', capitalizedWords: true }
    }

    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/info/69')
  })

  it('parse action into path with parameters using route object containing capitalizedWords: true: /info/foo-bar', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'Foo Bar' } }
    const routesMap = {
      INFO_PARAM: { path: '/info/:param', capitalizedWords: true }
    }

    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/info/foo-bar')
  })

  it('parse action into path with parameters using route object containing toPath() function: /info/foo-param-bar', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'Foo Bar' } }
    const routesMap = {
      INFO_PARAM: {
        path: '/info/:param',
        toPath: (value, key) => value.replace(' ', `-${key}-`).toLowerCase()
      }
    }

    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/info/foo-param-bar')
  })

  it('perform no formatting when route object contains ONLY path key: /info/FooBar', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'FooBar' } }
    const routesMap = {
      INFO_PARAM: { path: '/info/:param' }
    }

    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/info/FooBar')
  })

  it('throw error when parsing non-matched action', () => {
    const routesMap = {
      INFO: { path: '/info' }
    }

    let performMatch = () => actionToPath({ type: 'MISSED' }, routesMap)
    expect(performMatch).toThrowError()

    performMatch = () => actionToPath({ type: 'INFO' }, routesMap)
    expect(performMatch).not.toThrowError()
  })

  it('never returns an empty string when path has single optional param that is undefined', () => {
    const action = { type: 'INFO_PARAM', payload: { param: undefined } }
    const routesMap = { INFO_PARAM: '/:param?' }
    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/')
  })

  it('forwards toPathOptions to path-to-regexp toPath', () => {
    const action = { type: 'INFO', payload: { param: '1,2' } }
    const routesMap = {
      INFO: {
        path: '/info/:param',
        toPathOptions: {
          encode: param => param
        }
      }
    }
    const path = actionToPath(action, routesMap) /*? */
    expect(path).toEqual('/info/1,2')
  })
})

describe('changePageTitle()', () => {
  it('when title changes set it to document.title', () => {
    const document = {}
    const title = 'foo'

    const ret = changePageTitle(document, title)

    expect(document).toEqual({ title: 'foo' })
    expect(ret).toEqual('foo')
  })

  it('when title changes do not set document.title', () => {
    const document = { title: 'foo' }
    const title = 'foo'

    const ret = changePageTitle(document, title)

    expect(document).toEqual({ title: 'foo' })
    expect(ret).toEqual(null) // no return value when title does not change
  })
})
