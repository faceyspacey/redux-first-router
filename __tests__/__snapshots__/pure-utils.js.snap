// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nestAction(pathname, receivedAction, prevLocation, history, kind?) nestAction properly formats/nests action object 1`] = `
Object {
  "meta": Object {
    "info": "something",
    "location": Object {
      "current": Object {
        "pathname": "/path",
        "payload": Object {
          "bar": "baz",
        },
        "type": "FOO",
      },
      "history": undefined,
      "kind": undefined,
      "prev": Object {
        "pathname": "previous",
        "payload": Object {
          "bla": "prev",
        },
        "type": "PREV",
      },
    },
  },
  "payload": Object {
    "bar": "baz",
  },
  "type": "FOO",
}
`;

exports[`nestAction(pathname, receivedAction, prevLocation, history, kind?) nestHistory formats simplified history object for action + state 1`] = `
Object {
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/",
      "search": "",
      "state": undefined,
    },
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/foo",
      "search": "",
      "state": undefined,
    },
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/bar/baz",
      "search": "",
      "state": undefined,
    },
  ],
  "index": 2,
  "length": 3,
}
`;
